### ISSUE TITLE
App seems to stick but doesn't crash

### ISSUE THREAD

**Issue by undertheblu3sky** [shadowmoose/RedditDownloader#78]
### Describe the bug
The best I can give is that the app seems to stay at same progress after running so long, it could be a specific file but as of right now it's been stuck over an hour at deduplication (464) files with 2510 in queue and 8340 files completed. Only errors I see in cmd prompt is failed links or https failing which I'm guessing is from bad links.

### Environment Info
 - **OS:** Windows 10 1903
 - **RMD Version:** 3.0.0 (with whatever changes you made to main branch)

**Comment by shadowmoose** [comment]
Odd, I'll have to look into this. I haven't run into any issues with the deduplication process, but it's possible there's some file-related bug I have yet to run into during testing.

**Comment by undertheblu3sky** [comment]
Also I noticed I have the same issue as https://github.com/shadowmoose/RedditDownloader/issues/48
where I have files with multiple extensions. I deleted the sql databases and all downloaded files thinking maybe that could be an issue and I just stopped the run I stated earlier about 30 minutes ago.  so it ran about 6 hours. before closing with no duplicates. I'm going to restart it without cleaning any files and see if I get duplicate extensions.

**Comment by shadowmoose** [comment]
If RMD is saving to an existing, older directory with a missing database (or an older manifest from RMD 2.0) it is likely you will get duplicates. RMD only knows about the files it has personally saved to the db. 

The run-time deduplication bug is very odd, however. It should never hang for long. At the worst, RMD might take a few seconds to load (and hash) a massive file it's downloaded, but that should still complete about as fast as your disk can read the file back. My guess is that something's either crashing in the deduplication process and it's not detecting the failure, or perhaps it's getting stuck in a loop trying to access a file somehow. I may have to add more debugging to track this one down.

**Comment by undertheblu3sky** [comment]
So I left it on all night.  it's stuck again on 2506 files in queue. and I also now have the above issue. which had the same database that I closed and restarted with last night. it just has files with multiple extensions. IE .jpg.jpg. I wonder if when I close the app when it's stuck like this the database doesn't update what was downloaded or it hasn't had the chance to deduplicate and it doesn't recheck for duplicates.

**Comment by undertheblu3sky** [comment]
Finally got an error, Could this be related?

```
Process RedditElementLoader:
Traceback (most recent call last):
  File "C:\Program Files (x86)\Python37-32\lib\multiprocessing\process.py", line 297, in _bootstrap
    self.run()
  File "redditdownloader\processing\redditloader.py", line 30, in run
    self.load()
  File "redditdownloader\processing\redditloader.py", line 47, in load
    self._scan_sources()
  File "redditdownloader\processing\redditloader.py", line 73, in _scan_sources
    self._create_url_file(u, post=post)
  File "redditdownloader\processing\redditloader.py", line 114, in _create_url_file
    filename = name_generator.choose_file_name(url=url, post=post, session=self._session, album_size=album_size)
  File "redditdownloader\processing\name_generator.py", line 16, in choose_file_name
    base = nb % idx
TypeError: not enough arguments for format string
```

**Comment by shadowmoose** [comment]
Hey, I haven't been able to duplicate the freezing issue, but that stack trace you posted helped me spot another bug. Thanks for capturing that. 

As of 4aee7fdb40a0e78cae54e50e9c788c948909d206, that formatting TypeError should be safely avoided.

**Comment by undertheblu3sky** [comment]
I actually think that was the whole problem I'm going to test with your latest commit. I think it wasn't reporting it. What I meant by freeze was the progress would just sit there. the app wouldn't actually crash or not respond

**Comment by undertheblu3sky** [comment]
https://i.imgur.com/TWN9oT5.png

It's just sits here. I'm not sure if it's a file. The last message in cmd prompt is Deleted Orphan Files.

**Comment by undertheblu3sky** [comment]
update: left it on all night and this is what it looks like now. https://i.imgur.com/g67lsIi.png files were added to queue but it seems stuck on the file it's trying to deal with. the last thing in Cmd prompt is deleted orphan files.

**Comment by shadowmoose** [comment]
Hmm, in your first image, the "orange" color of the Threads is consistent with them sitting at idle, not having target URLs to download - which doesn't make sense, given that the queue is still actively loading Posts. In fact, it appears as though the queue is full, and RMD is stuck waiting for some Posts to be downloaded before it loads more.

In the second screenshot it appears that deduplication is disabled, but you're still getting logs from the deduplication process - which also shouldn't happen. The second screenshot appears to have progressed quite a bit, but if you're not seeing any progress between the downloading threads or the queue count, that is an issue.

I'm going to have to add some debugging to hunt this down, or somehow replicated it on my setups. My best guess is that there's some issue with communication between the subprocesses or database, which is either slowing them down or disabling them entirely.

If you're interested, I've just pushed a new commit (9a94e602d4c041d13becd41b206b09e716843ff4) which should more easily display the status of each downloading process. In the event a failure is encountered by a downloader which requires it to exit, its status will display the error message.

Otherwise, I'll look at ways to stress-test this more soon - once I have a bit more time.

**Comment by undertheblu3sky** [comment]
this is a big error I edited it remove my username in windows and which reddituser i was downloading the subreddit and the title of the name. I'm going to try a new renaming system to see if it works. but this is the only type of error I'm getting

`(sqlite3.OperationalError) database is locked
[SQL: UPDATE files SET path=?, downloaded=? WHERE files.id = ?]
[parameters: ('Reddituser/subreddit/title.png', 1, 14674)]
(Background on this error at: http://sqlalche.me/e/e3q8)(sqlite3.OperationalError) database is locked
[SQL: UPDATE files SET path=?, downloaded=? WHERE files.id = ?]
[parameters: ('Reddituser/subreddit/title.png', 1, 14672)]
(Background on this error at: http://sqlalche.me/e/e3q8)

Traceback (most recent call last):
Traceback (most recent call last):
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\engine\base.py", line 1249, in _execute_context
    cursor, statement, parameters, context
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\engine\default.py", line 580, in do_execute
    cursor.execute(statement, parameters)
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\engine\base.py", line 1249, in _execute_context
    cursor, statement, parameters, context
sqlite3.OperationalError: database is locked
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\engine\default.py", line 580, in do_execute
    cursor.execute(statement, parameters)

The above exception was the direct cause of the following exception:

sqlite3.OperationalError: database is locked
Traceback (most recent call last):

The above exception was the direct cause of the following exception:

  File "redditdownloader\processing\downloader.py", line 68, in run
    self._session.commit()
Traceback (most recent call last):
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\orm\session.py", line 1027, in commit
    self.transaction.commit()
  File "redditdownloader\processing\downloader.py", line 68, in run
    self._session.commit()
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\orm\session.py", line 494, in commit
    self._prepare_impl()
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\orm\session.py", line 1027, in commit
    self.transaction.commit()
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\orm\session.py", line 473, in _prepare_impl
    self.session.flush()
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\orm\session.py", line 494, in commit
    self._prepare_impl()
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\orm\session.py", line 2470, in flush
    self._flush(objects)
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\orm\session.py", line 473, in _prepare_impl
    self.session.flush()
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\orm\session.py", line 2608, in _flush
    transaction.rollback(_capture_exception=True)
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\orm\session.py", line 2470, in flush
    self._flush(objects)
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\util\langhelpers.py", line 68, in __exit__
    compat.reraise(exc_type, exc_value, exc_tb)
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\orm\session.py", line 2608, in _flush
    transaction.rollback(_capture_exception=True)
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\util\compat.py", line 153, in reraise
    raise value
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\util\langhelpers.py", line 68, in __exit__
    compat.reraise(exc_type, exc_value, exc_tb)
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\orm\session.py", line 2568, in _flush
    flush_context.execute()
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\util\compat.py", line 153, in reraise
    raise value
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\orm\unitofwork.py", line 422, in execute
    rec.execute(self)
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\orm\session.py", line 2568, in _flush
    flush_context.execute()
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\orm\unitofwork.py", line 589, in execute
    uow,
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\orm\unitofwork.py", line 422, in execute
    rec.execute(self)
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\orm\persistence.py", line 236, in save_obj
    update,
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\orm\unitofwork.py", line 589, in execute
    uow,
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\orm\persistence.py", line 996, in _emit_update_statements
    statement, multiparams
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\orm\persistence.py", line 236, in save_obj
    update,
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\engine\base.py", line 988, in execute
    return meth(self, multiparams, params)
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\orm\persistence.py", line 996, in _emit_update_statements
    statement, multiparams
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\sql\elements.py", line 287, in _execute_on_connection
    return connection._execute_clauseelement(self, multiparams, params)
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\engine\base.py", line 988, in execute
    return meth(self, multiparams, params)
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\engine\base.py", line 1107, in _execute_clauseelement
    distilled_params,
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\sql\elements.py", line 287, in _execute_on_connection
    return connection._execute_clauseelement(self, multiparams, params)
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\engine\base.py", line 1253, in _execute_context
    e, statement, parameters, cursor, context
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\engine\base.py", line 1107, in _execute_clauseelement
    distilled_params,
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\engine\base.py", line 1473, in _handle_dbapi_exception
    util.raise_from_cause(sqlalchemy_exception, exc_info)
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\engine\base.py", line 1253, in _execute_context
    e, statement, parameters, cursor, context
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\util\compat.py", line 398, in raise_from_cause
    reraise(type(exception), exception, tb=exc_tb, cause=cause)
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\engine\base.py", line 1473, in _handle_dbapi_exception
    util.raise_from_cause(sqlalchemy_exception, exc_info)
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\util\compat.py", line 152, in reraise
    raise value.with_traceback(tb)
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\util\compat.py", line 398, in raise_from_cause
    reraise(type(exception), exception, tb=exc_tb, cause=cause)
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\engine\base.py", line 1249, in _execute_context
    cursor, statement, parameters, context
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\util\compat.py", line 152, in reraise
    raise value.with_traceback(tb)
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\engine\default.py", line 580, in do_execute
    cursor.execute(statement, parameters)
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\engine\base.py", line 1249, in _execute_context
    cursor, statement, parameters, context
sqlalchemy.exc.OperationalError: (sqlite3.OperationalError) database is locked
[SQL: UPDATE files SET path=?, downloaded=? WHERE files.id = ?]
[parameters: ('Reddituser/subreddit/title.png', 1, 14674)]
(Background on this error at: http://sqlalche.me/e/e3q8)
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\engine\default.py", line 580, in do_execute
    cursor.execute(statement, parameters)
sqlalchemy.exc.OperationalError: (sqlite3.OperationalError) database is locked
[SQL: UPDATE files SET path=?, downloaded=? WHERE files.id = ?]
[parameters: ('Reddituser/subreddit/title.png', 1, 14672)]
(Background on this error at: http://sqlalche.me/e/e3q8)
(sqlite3.OperationalError) database is locked
[SQL: UPDATE files SET path=?, downloaded=? WHERE files.id = ?]
[parameters: ('Reddituser/subreddit/title.png', 1, 14673)]
(Background on this error at: http://sqlalche.me/e/e3q8)
Traceback (most recent call last):
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\engine\base.py", line 1249, in _execute_context
    cursor, statement, parameters, context
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\engine\default.py", line 580, in do_execute
    cursor.execute(statement, parameters)
sqlite3.OperationalError: database is locked

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "redditdownloader\processing\downloader.py", line 68, in run
    self._session.commit()
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\orm\session.py", line 1027, in commit
    self.transaction.commit()
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\orm\session.py", line 494, in commit
    self._prepare_impl()
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\orm\session.py", line 473, in _prepare_impl
    self.session.flush()
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\orm\session.py", line 2470, in flush
    self._flush(objects)
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\orm\session.py", line 2608, in _flush
    transaction.rollback(_capture_exception=True)
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\util\langhelpers.py", line 68, in __exit__
    compat.reraise(exc_type, exc_value, exc_tb)
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\util\compat.py", line 153, in reraise
    raise value
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\orm\session.py", line 2568, in _flush
    flush_context.execute()
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\orm\unitofwork.py", line 422, in execute
    rec.execute(self)
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\orm\unitofwork.py", line 589, in execute
    uow,
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\orm\persistence.py", line 236, in save_obj
    update,
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\orm\persistence.py", line 996, in _emit_update_statements
    statement, multiparams
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\engine\base.py", line 988, in execute
    return meth(self, multiparams, params)
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\sql\elements.py", line 287, in _execute_on_connection
    return connection._execute_clauseelement(self, multiparams, params)
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\engine\base.py", line 1107, in _execute_clauseelement
    distilled_params,
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\engine\base.py", line 1253, in _execute_context
    e, statement, parameters, cursor, context
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\engine\base.py", line 1473, in _handle_dbapi_exception
    util.raise_from_cause(sqlalchemy_exception, exc_info)
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\util\compat.py", line 398, in raise_from_cause
    reraise(type(exception), exception, tb=exc_tb, cause=cause)
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\util\compat.py", line 152, in reraise
    raise value.with_traceback(tb)
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\engine\base.py", line 1249, in _execute_context
    cursor, statement, parameters, context
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\engine\default.py", line 580, in do_execute
    cursor.execute(statement, parameters)
sqlalchemy.exc.OperationalError: (sqlite3.OperationalError) database is locked
[SQL: UPDATE files SET path=?, downloaded=? WHERE files.id = ?]
[parameters: ('Reddituser/subreddit/title.png', 1, 14673)]
(Background on this error at: http://sqlalche.me/e/e3q8)
(sqlite3.OperationalError) database is locked
[SQL: UPDATE files SET path=?, downloaded=? WHERE files.id = ?]
[parameters: ('reddituser2/subreddit2/Hope everyone had a great holiday!/1.jpg', 1, 14661)]
(Background on this error at: http://sqlalche.me/e/e3q8)
Traceback (most recent call last):
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\engine\base.py", line 1249, in _execute_context
    cursor, statement, parameters, context
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\engine\default.py", line 580, in do_execute
    cursor.execute(statement, parameters)
sqlite3.OperationalError: database is locked

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "redditdownloader\processing\downloader.py", line 68, in run
    self._session.commit()
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\orm\session.py", line 1027, in commit
    self.transaction.commit()
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\orm\session.py", line 494, in commit
    self._prepare_impl()
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\orm\session.py", line 473, in _prepare_impl
    self.session.flush()
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\orm\session.py", line 2470, in flush
    self._flush(objects)
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\orm\session.py", line 2608, in _flush
    transaction.rollback(_capture_exception=True)
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\util\langhelpers.py", line 68, in __exit__
    compat.reraise(exc_type, exc_value, exc_tb)
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\util\compat.py", line 153, in reraise
    raise value
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\orm\session.py", line 2568, in _flush
    flush_context.execute()
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\orm\unitofwork.py", line 422, in execute
    rec.execute(self)
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\orm\unitofwork.py", line 589, in execute
    uow,
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\orm\persistence.py", line 236, in save_obj
    update,
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\orm\persistence.py", line 996, in _emit_update_statements
    statement, multiparams
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\engine\base.py", line 988, in execute
    return meth(self, multiparams, params)
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\sql\elements.py", line 287, in _execute_on_connection
    return connection._execute_clauseelement(self, multiparams, params)
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\engine\base.py", line 1107, in _execute_clauseelement
    distilled_params,
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\engine\base.py", line 1253, in _execute_context
    e, statement, parameters, cursor, context
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\engine\base.py", line 1473, in _handle_dbapi_exception
    util.raise_from_cause(sqlalchemy_exception, exc_info)
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\util\compat.py", line 398, in raise_from_cause
    reraise(type(exception), exception, tb=exc_tb, cause=cause)
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\util\compat.py", line 152, in reraise
    raise value.with_traceback(tb)
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\engine\base.py", line 1249, in _execute_context
    cursor, statement, parameters, context
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\engine\default.py", line 580, in do_execute
    cursor.execute(statement, parameters)
sqlalchemy.exc.OperationalError: (sqlite3.OperationalError) database is locked
[SQL: UPDATE files SET path=?, downloaded=? WHERE files.id = ?]
[parameters: ('reddituser2/subreddit2/Hope everyone had a great holiday!/1.jpg', 1, 14661)]
(Background on this error at: http://sqlalche.me/e/e3q8)
(sqlite3.OperationalError) database is locked
[SQL: UPDATE files SET path=?, downloaded=? WHERE files.id = ?]
[parameters: ('Reddituser/subreddit/title.png', 1, 14668)]
(Background on this error at: http://sqlalche.me/e/e3q8)
Traceback (most recent call last):
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\engine\base.py", line 1249, in _execute_context
    cursor, statement, parameters, context
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\engine\default.py", line 580, in do_execute
    cursor.execute(statement, parameters)
sqlite3.OperationalError: database is locked

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "redditdownloader\processing\downloader.py", line 68, in run
    self._session.commit()
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\orm\session.py", line 1027, in commit
    self.transaction.commit()
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\orm\session.py", line 494, in commit
    self._prepare_impl()
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\orm\session.py", line 473, in _prepare_impl
    self.session.flush()
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\orm\session.py", line 2470, in flush
    self._flush(objects)
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\orm\session.py", line 2608, in _flush
    transaction.rollback(_capture_exception=True)
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\util\langhelpers.py", line 68, in __exit__
    compat.reraise(exc_type, exc_value, exc_tb)
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\util\compat.py", line 153, in reraise
    raise value
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\orm\session.py", line 2568, in _flush
    flush_context.execute()
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\orm\unitofwork.py", line 422, in execute
    rec.execute(self)
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\orm\unitofwork.py", line 589, in execute
    uow,
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\orm\persistence.py", line 236, in save_obj
    update,
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\orm\persistence.py", line 996, in _emit_update_statements
    statement, multiparams
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\engine\base.py", line 988, in execute
    return meth(self, multiparams, params)
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\sql\elements.py", line 287, in _execute_on_connection
    return connection._execute_clauseelement(self, multiparams, params)
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\engine\base.py", line 1107, in _execute_clauseelement
    distilled_params,
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\engine\base.py", line 1253, in _execute_context
    e, statement, parameters, cursor, context
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\engine\base.py", line 1473, in _handle_dbapi_exception
    util.raise_from_cause(sqlalchemy_exception, exc_info)
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\util\compat.py", line 398, in raise_from_cause
    reraise(type(exception), exception, tb=exc_tb, cause=cause)
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\util\compat.py", line 152, in reraise
    raise value.with_traceback(tb)
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\engine\base.py", line 1249, in _execute_context
    cursor, statement, parameters, context
  File "C:\Users\windowsuser\AppData\Roaming\Python\Python37\site-packages\sqlalchemy\engine\default.py", line 580, in do_execute
    cursor.execute(statement, parameters)
sqlalchemy.exc.OperationalError: (sqlite3.OperationalError) database is locked
[SQL: UPDATE files SET path=?, downloaded=? WHERE files.id = ?]
[parameters: ('Reddituser/subreddit/title.png', 1, 14668)]
(Background on this error at: http://sqlalche.me/e/e3q8)`

**Comment by shadowmoose** [comment]
Interesting, it looks like the downloaders are all failing because they cannot save to the sqlite database in a timely manner. Either the disk (or queries) somehow can't keep up and the database is staying locked for too long, or the database file itself is locked.

I don't suppose there are any background processes still running, that may be using the database? If not, you could also open the database file with a program like https://github.com/sqlitebrowser/sqlitebrowser, and see if you're able to browse/save any changes.

**Comment by undertheblu3sky** [comment]
I was able to write to the database after having it run all night with latest commit. Maybe it needs to loop to recheck to finish? or check if database is locked and wait till it's writable?

**Comment by undertheblu3sky** [comment]
Was looking at possible fixes for locked database. found this stackoverflow cursor.close() asap after query(even read-onlys) https://stackoverflow.com/a/53470179

**Comment by shadowmoose** [comment]
This is a difficult one to hunt down. The database is wrapped in an ORM library, SQLAlchemy, which handles most of the session termination itself. Additionally, RMD generally closes its sessions as soon as possible to avoid these types of problems.

If it was a simple case of a session locking the database for too long after forgetting to close, this should manifest in any of the tests RMD runs. As it stands, I'm trying to hunt down anywhere RMD could possibly be holding a lock for too long. I'm currently unaware of any problem areas.

My best guess is it's either related to a crash (which should've logged an error somewhere, so this is less likely), or something in a subprocess is taking too long to complete, and the other connections are timing out. If you have it enabled, deduplication may be the culprit.

Do you have file deduplication enabled? If so, are you still seeing the ```Deleted orphan Files``` line after the errors?

**Comment by undertheblu3sky** [comment]
I do have depublication enabled. I tried a fresh browser thinking maybe it's addons which I'm using firefox and this happened https://i.imgur.com/OWT7EWu.png . I'll try it without deduplication if it stops with all threads on database error next.

**Comment by undertheblu3sky** [comment]
`Process RedditElementLoader:
Traceback (most recent call last):
  File "C:\Program Files (x86)\Python37-32\lib\multiprocessing\process.py", line 297, in _bootstrap
    self.run()
  File "redditdownloader\processing\redditloader.py", line 30, in run
    self.load()
  File "redditdownloader\processing\redditloader.py", line 47, in load
    self._scan_sources()
  File "redditdownloader\processing\redditloader.py", line 61, in _scan_sources
    for r in source.get_elements():
  File "redditdownloader\sources\pushshift_user_posts.py", line 25, in get_elements
    sub = list(ps.search_submissions(ids=post.link_id.replace('t3_', '', 1), limit=1))[0]
IndexError: list index out of range`

Yay a new error. this happened without dedupe on.

**Comment by shadowmoose** [comment]
I see. That error looks to be caused by Pushshift sending RMD a Submission ID that Pushshift also doesn't have any information about. 

I didn't realize that was possible. I'll add some error handling to catch those User Comment issues.

**Comment by undertheblu3sky** [comment]
So the database lock happened again with dedupe's off.  :( is it possible I have too many sources? I made a source for every user I wanted to download their posts. That link I posted of stackoverflow said sometimes it's a multiple thread issue. and that's why the guy suggests to use the cursor.close() whenever possible. Maybe adding that at the end of the wrappers?

**Comment by shadowmoose** [comment]
Resolved as of [Release 3.0.0](https://github.com/shadowmoose/RedditDownloader/releases/tag/3.0.0). Feel free to reopen if this still needs more attention.
